### PR TITLE
Redo JSitor

### LINKED ISSUE #4734: Allow JSitor Tag to accept Full URL and parameters
Continuing the conversation from #4657, we'd like the JSitor Liquid tag to be able to accept the full URL, as well as any additional URL parameters. For example: `https://jsitor.com/embed/1QgJVmCam?html&js&css&result&light&consoleOnly&`

We'll need to extract the ID from that link, as well as the `html`, `js`, `css`, `result`, `light` and `consoleOnly` parameters. Once extracted, the HTML file should have those parameters inserted back in. Alternatively, we can simply validate the given URL and its parameters and place it back in, but we'll need to watch out for any sanitization that might happen in the process.

I think we'll want the final use case look like this:
```
{% jsitor https://jsitor.com/embed/1QgJVmCam?html&js&css&result&light&consoleOnly& %}
#=> renders the liquid tag properly
```

### PR DESCRIPTION
**Pull request by artoodeeto**
- can now accept full url with params
- added more test
- redo jsitor guide

## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
as per request. Jsitor should now accept full url with optional parameters
## Related Tickets & Documents
resolves #4734 
## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
dev

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by artoodeeto** [comment]
@ashvin777 @Zhao-Andy @rhymes hello guys. sorry took too long. here you go. let me know if I miss something. :)

**Comment by ashvin777** [comment]
Thanks @artoodeeto . 

3 Requests -
1. Can you change the order of this tag in documentation? Add it under "JSFiddle Embed"? I am asking because this tool is somewhat similar to JSFiddle/CodePen so if this can be added below it then it will be great.
3. Can you reduce the height of embed view to 400px? 600px seems too big to me and takes too much space on the page.
4. Also if you can add some more documentation about custom tabs then it will be great. Example - Check "JSFiddle Embed" documentation. If already added please ignore.

**Comment by Zhao-Andy** [comment]
Agree with @rhymes about supporting both the link and the ID. Sorry if that wasn't clear!

**Comment by artoodeeto** [comment]
@ashvin777 @rhymes @Zhao-Andy hello guys. yea for sure. will work on this tonight. :)

**Comment by artoodeeto** [comment]
@rhymes @Zhao-Andy question. are we going to allow them to include the parameters if they only pass the ID? or only accept the ID. what do you think?

**Comment by Zhao-Andy** [comment]
@artoodeeto we'd want to allow the ID and parameters, so maybe a final result would be something like:
```
{% jsitor idC0D3e?html?js?other?parameters %}
# or 
{% jsitor idC0D3e %}
# or
{% jsitor https://full-url.com/embed/idC0D3e?html?js?other?parameters %}
# or 
{% jsitor https://full-url.com/embed/idC0D3e %}
```

**Comment by artoodeeto** [comment]
hey guys the test is failing but its not on the area that im working.  It happened when I updated my repo. let me try to fix this.

**Comment by artoodeeto** [comment]
oh crap I think I caused this. you would have caught this if the test failed but I have a different error. its about the listing. `user_views_a_reading_list_spec.rb:14`

**Comment by artoodeeto** [comment]
I also have warnings on my test, do you think this is the caused? im gonna try to fix this. sorry. ✌️

File: app/liquid_tags/jsitor_tag.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 class JsitorTag < LiquidTagBase
   PARTIAL = "liquids/jsitor".freeze
   URL_REGEXP = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)([?a-zA-Z&]*)\Z/.freeze
+  ID_REGEXP = /\A([?a-zA-Z0-9&])*\Z/.freeze
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: I think this regexp is unusually complicated. 

It seems to me what you're generally trying to match is every ASCII char and every digit in addition to the `&` and `?` chars, with no specific length of the whole string. You also don't need grouping.

I think it can be simplified to `\A[\w&?]+\Z` as any word character, plus the `&`, `?`, repeated once or more.

I tested them with Rubular

What do you think?

File: app/liquid_tags/jsitor_tag.rb [hunk 2]
```diff
@@ -21,18 +22,21 @@ def render(_context)
...
   def jsitor_link_parser(link)
     parsed_link = ActionController::Base.helpers.strip_tags(link.strip).gsub("amp;", "")
-
-    return parsed_link if link_valid?(parsed_link)
-
-    jsitor_error
+    link_valid?(parsed_link)
   end
 
   def link_valid?(link)
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: this should probably be called `validate_link` since it raises an exception.

Otherwise you can have `jsitor_link_parser` do `link_valid?() || jsitor_error`

File: app/liquid_tags/jsitor_tag.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 class JsitorTag < LiquidTagBase
   PARTIAL = "liquids/jsitor".freeze
   URL_REGEXP = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)([?a-zA-Z&]*)\Z/.freeze
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: I think this can be simplified to `\A(https|http):\/\/jsitor\.com\/embed\/\w+[?a-zA-Z&]*\Z`

you don't need grouping since you're just passing the link verbatim to the HTML

File: app/liquid_tags/jsitor_tag.rb [hunk 4]
```diff
@@ -21,18 +22,21 @@ def render(_context)
...
-    URL_REGEXP.match link
+    return link if URL_REGEXP.match link
+    return "https://jsitor.com/embed/#{link}" if ID_REGEXP.match link
+
+    jsitor_error
   end
 
   def jsitor_error
-    raise StandardError, "Invalid JSitor link. Link should have /embed/. ex: https://jsitor.com/embed/1QgJVmCam. Parameters are optional"
+    raise StandardError, "Invalid JSitor link. Link should have /embed/.
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: I don't think we need to include the whole explanation. 

Maybe we can just say: `Invalid JSitor link. Please read the editor guide for more information`

File: app/liquid_tags/jsitor_tag.rb (hunk 1 above)
Comment (artoodeeto) [reply app/liquid_tags/jsitor_tag.rb]: ohh thats cool. yea for sure I didnt know that. thank you. also if I dont group it for some reason it wont validate I tried it on ruby console this part took me long time to figure out because of not grouping it but let me try again.

File: app/liquid_tags/jsitor_tag.rb (hunk 2 above)
Comment (artoodeeto) [reply app/liquid_tags/jsitor_tag.rb]: ill just change the method name. :smile: `validate_link`

File: app/liquid_tags/jsitor_tag.rb (hunk 4 above)
Comment (artoodeeto) [reply app/liquid_tags/jsitor_tag.rb]: I knew it. I was thinking to add a link to the guide on error too. haha. ill change this one.

**Comment by artoodeeto** [comment]
@rhymes btw what was the error all about? didn't finish last night was so sleepy. did I cause the problem?

**Comment by rhymes** [comment]
it was a flaky test, I restarted the build

**Comment by artoodeeto** [comment]
@ashvin777  hello check these changes. let me know if I could fix anything. :)

**Comment by artoodeeto** [comment]
@rhymes yoshh! Thank you! No problem I can wait. :smile:

**Comment by ashvin777** [comment]
Awesome work @artoodeeto . And thank you @rhymes @Zhao-Andy and @benhalpern for your 
reviews. 

Its looks super awesome - 
image

**Comment by ashvin777** [comment]
Found a small issue - Raised  https://github.com/thepracticaldev/dev.to/issues/4823

It is not accepting snippet's id having `-` in between. Its a valid snippet id - https://jsitor.com/2o-syYxmi

Screenshot 2019-11-15 at 1 03 02 AM

**Comment by artoodeeto** [comment]
ahhh. github doesnt show on recent activity. sorry i didn't notice. :cry: @ashvin777
